src/solvers/strings/string_dependencies.cpp

```diff
@@ -0,0 +1,363 @@
+  const std::function<void(const array_string_exprt &)> f);
+
+/// Construct a string_builtin_functiont object from a function application
+/// \return a unique pointer to the created object
+static std::unique_ptr<string_builtin_functiont> to_string_builtin_function(
+  const function_application_exprt &fun_app,
+  const exprt &return_code,
+  array_poolt &array_pool)
+{
+  const auto name = expr_checked_cast<symbol_exprt>(fun_app.function());
+  PRECONDITION(!is_ssa_expr(name));
+
+  const irep_idt &id = name.get_identifier();
+
+  if(id == ID_cprover_string_insert_func)
+    return util_make_unique<string_insertion_builtin_functiont>(
+      return_code, fun_app.arguments(), array_pool);
+
+  if(id == ID_cprover_string_concat_func)
+    return util_make_unique<string_concatenation_builtin_functiont>(
+      return_code, fun_app.arguments(), array_pool);
+
+  if(id == ID_cprover_string_concat_char_func)
+    return util_make_unique<string_concat_char_builtin_functiont>(
+      return_code, fun_app.arguments(), array_pool);
+
+  if(id == ID_cprover_string_of_int_func)
+    return util_make_unique<string_of_int_builtin_functiont>(
+      return_code, fun_app.arguments(), array_pool);
+
+  if(id == ID_cprover_string_char_set_func)
+    return util_make_unique<string_set_char_builtin_functiont>(
+      return_code, fun_app.arguments(), array_pool);
+
+  if(id == ID_cprover_string_to_lower_case_func)
+    return util_make_unique<string_to_lower_case_builtin_functiont>(
+      return_code, fun_app.arguments(), array_pool);
+
+  if(id == ID_cprover_string_to_upper_case_func)
+    return util_make_unique<string_to_upper_case_builtin_functiont>(
+      return_code, fun_app.arguments(), array_pool);
+
+  return util_make_unique<string_builtin_function_with_no_evalt>(
+    return_code, fun_app, array_pool);
+}
+
+string_dependenciest::string_nodet &
+string_dependenciest::get_node(const array_string_exprt &e)
+{
+  auto entry_inserted = node_index_pool.emplace(e, string_nodes.size());
+  if(!entry_inserted.second)
+    return string_nodes[entry_inserted.first->second];
+
+  string_nodes.emplace_back(e, entry_inserted.first->second);
+  return string_nodes.back();
+}
+
+std::unique_ptr<const string_dependenciest::string_nodet>
+string_dependenciest::node_at(const array_string_exprt &e) const
+{
+  const auto &it = node_index_pool.find(e);
+  if(it != node_index_pool.end())
+    return util_make_unique<const string_nodet>(string_nodes.at(it->second));
+  return {};
+}
+
+string_dependenciest::builtin_function_nodet &string_dependenciest::make_node(
+  std::unique_ptr<string_builtin_functiont> &builtin_function)
+{
+  builtin_function_nodes.emplace_back(
+    std::move(builtin_function), builtin_function_nodes.size());
+  return builtin_function_nodes.back();
+}
+
+const string_builtin_functiont &string_dependenciest::get_builtin_function(
+  const builtin_function_nodet &node) const
+{
+  return *node.data;
+}
+
+static void for_each_atomic_string(
```

Comment:
:question: Is this function related only to dependencies?

Reply:
Probably not, but it's not used anywhere else yet.